source/dem/plane_insertion.cc

```diff
@@ -0,0 +1,338 @@
+                {
+                  // At this point, we know the current processor have to many
+                  // available cell to insert in (or just the right amount). The
+                  // excess number of available cell on this processor is equal
+                  // to the positive value of the
+                  // remained_particles_of_each_type variable since that
+                  // variable became negative on this processor.
+
+                  // Thus, we decrease the number of particle to insert on the
+                  // current processor. To do so, we add the negative variable
+                  // to the number of available cells which guarantee that the
+                  // number of inserted particles in the simulation will reach
+                  // its maximum on this processor.
+
+                  number_of_particles_to_insert_per_core[i] +=
+                    remained_particles_of_each_type;
+
+                  // We put the remained_particles_of_each_type variable  back
+                  // to zero since we have just lower the number of cell
+                  // available on this processor by the exceeding amount.
+                  remained_particles_of_each_type = 0;
+
+                  starting_IDs_on_every_proc[i] = starting_id_on_proc;
+                  starting_id_on_proc +=
+                    number_of_particles_to_insert_per_core[i];
+
+                  // Loop over the remaining processor and put their number of
+                  // available cell to zero.
+                  for (unsigned int j = i + 1; j < (total_mpi_process); ++j)
+                    {
+                      number_of_particles_to_insert_per_core[j] = 0;
+                      starting_IDs_on_every_proc[j]             = 0;
+                    }
+                  break;
+                }
+              starting_IDs_on_every_proc[i] = starting_id_on_proc;
+              starting_id_on_proc += number_of_particles_to_insert_per_core[i];
+            }
+          std::fill(remained_particles_for_every_proc.begin(),
+                    remained_particles_for_every_proc.end(),
+                    remained_particles_of_each_type);
+        }
+
+      // Now, processor zero knows how many particles will be inserted by every
+      // processor without exceeding the maximum number of particles.
+
+      // We now need to scatter 3 information processor, first how
+      // many particles it needs to insert. Second what is the ID of the
+      // first particle it will insert at this insertion step. The starting ID
+      // is mandatory, since we don't want two particles to have the same ID,
+      // and we don't want to use the global_insertion method because it uses an
+      // all_gather, which has a high computational cost. Third, the
+      // remained_particles_of_each_type variable need to be updated proc other
+      // than proc 0 will enter the insert function dans the simulation will
+      // stop.
```

Comment:
```suggestion
      // all_gather, which has a high computational cost. Third, the
      // remained_particles_of_each_type variable needs to be updated on processors other
      // than processor 0 otherwise the insert function will become incoherent.
```